src/main/java/com/wire/bots/roman/MessageHandler.java

```diff
@@ -111,116 +124,120 @@ public void onReaction(WireClient client, ReactionMessage msg) {
-//
-//            message.attachment = Base64.getEncoder().encodeToString(img);
-            message.text = msg.getName();
-            message.mimeType = msg.getMimeType();
-            message.size = msg.getSize();
+            OutgoingMessage message = createOutgoingMessage(botId, type, client.getConversationId(), msg);
+            message.attachment = new Attachment();
+            message.attachment.name = msg.getName();
+            message.attachment.mimeType = msg.getMimeType();
+            message.attachment.size = msg.getSize();
 
-            message.meta = extractAssetMeta(msg);
+            outgoingMessageDAO.insert(message.messageId, mapper.writeValueAsString(message));
 
             send(message);
         } catch (Exception e) {
-            Logger.exception("onAttachment: %s", e, e.getMessage());
+            Logger.exception("onFilePreview: %s", e, e.getMessage());
         }
     }
 
     @Override
-    public void onAudio(WireClient client, AudioMessage msg) {
-        final String type = "conversation.audio.new";
+    public void onAudioPreview(WireClient client, AudioPreviewMessage msg) {
+        final String type = "conversation.audio.preview";
 
         UUID botId = client.getId();
 
         validate(botId);
 
         try {
-            OutgoingMessage message = getOutgoingMessage(botId, type, msg);
-            message.conversationId = client.getConversationId();
+            OutgoingMessage message = createOutgoingMessage(botId, type, client.getConversationId(), msg);
+            message.attachment = new Attachment();
+            message.attachment.name = msg.getName();
+            message.attachment.mimeType = msg.getMimeType();
+            message.attachment.size = msg.getSize();
+            message.attachment.duration = msg.getDuration();
+            message.attachment.levels = msg.getLevels();
 
-//            byte[] img = client.downloadAsset(msg.getAssetKey(),
-//                    msg.getAssetToken(),
-//                    msg.getSha256(),
-//                    msg.getOtrKey());
-//            message.attachment = Base64.getEncoder().encodeToString(img);
+            outgoingMessageDAO.insert(message.messageId, mapper.writeValueAsString(message));
 
-            message.text = msg.getName();
-            message.mimeType = msg.getMimeType();
-            message.size = msg.getSize();
-            message.duration = msg.getDuration();
-            message.levels = msg.getLevels();
+            send(message);
+        } catch (Exception e) {
+            Logger.exception("onAudioPreview: %s", e, e.getMessage());
+        }
+    }
+
+    @Override
+    public void onAssetData(WireClient client, RemoteMessage msg) {
+        final String type = "conversation.asset.data";
 
-            message.meta = extractAssetMeta(msg);
+        UUID botId = client.getId();
+
+        validate(botId);
+
+        try {
+            OutgoingMessage message = outgoingMessageDAO.get(msg.getMessageId());
```

Comment:
what if there's no preview? do all assets have preview? also what if Roman fails to write the preview to the database? then this ends on the NPE

I'm not exactly sure whether this is a valid point or not or how to handle it, I just wanted to highlight it 🙂

Reply:
the same applies for audio/pictures

Reply:
Thanks. I throw an exception now. At least it has some info what went wrong instead of good ol' NPE

Reply:
sounds good 👍